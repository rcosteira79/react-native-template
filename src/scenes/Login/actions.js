import * as session from '../../services/session';
import Locales from '../../resources/locales';

import { EMAIL_CHANGED, PASSWORD_CHANGED, LOGIN, LOGIN_FAIL, LOGIN_SUCCESS } from './actionTypes';
import { navigateToMain } from '../../navigation/actions';
import { emailRegex } from '../../services/helpers/index';

export const emailChanged = email => {
  return {
    type: EMAIL_CHANGED,
    payload: {
      email,
      validEmail: emailRegex.test(email),
      error: ''
    }
  };
};

export const passwordChanged = password => {
  return {
    type: PASSWORD_CHANGED,
    payload: {
      password,
      validPassword: password !== '',
      error: ''
    }
  };
};

export const loginUser = (email, password) => {
  return dispatch => {
    dispatch({ type: LOGIN });
    return session
      .authenticate(email, password)
      .then(() => {
        dispatch(loginSuccess());
        dispatch(navigateToMain());
      })
      .catch(error => {
        dispatch(loginFail(error));
      });
  };
};

const loginSuccess = () => {
  return {
    type: LOGIN_SUCCESS
  };
};

const loginFail = error => {
  const errorTextKey = getErrorMessage(error);

  return {
    type: LOGIN_FAIL,
    error: errorTextKey
  };
};

const getErrorMessage = error => {
  let errorTextKey;

  switch (error.response.status) {
    case 401:
      errorTextKey = Locales.login.error_invalid_email_password;
      break;

    default:
      errorTextKey = Locales.login.error_unknown;
      break;
  }

  return errorTextKey;
};
