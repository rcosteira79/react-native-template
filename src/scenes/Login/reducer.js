import {
  EMAIL_CHANGED,
  PASSWORD_CHANGED,
  LOGIN,
  LOGIN_SUCCESS,
  LOGIN_FAIL
} from './actionTypes';

export const initialState = {
  isLoading: false,
  email: '',
  password: '',
  validEmail: false,
  validPassword: false,
  error: ''
};

export const loginReducer = (state = initialState, action) => {
  switch (action.type) {
    case EMAIL_CHANGED:
    case PASSWORD_CHANGED:
      return { ...state, ...action.payload };
    case LOGIN:
      return { ...state, isLoading: true };
    case LOGIN_SUCCESS:
      return { ...state, ...initialState };
    case LOGIN_FAIL:
      return { ...state, error: action.error, isLoading: false };
    default:
      return state;
  }
};
