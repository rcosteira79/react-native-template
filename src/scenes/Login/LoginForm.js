import dismissKeyboard from 'react-native/Libraries/Utilities/dismissKeyboard';
import React from 'react';
import EStyleSheet from 'react-native-extended-stylesheet';
import { connect } from 'react-redux';
import { getTranslate } from 'react-localize-redux';
import { hook, wrap } from 'cavy';

import {
  View,
  TouchableOpacity,
  TouchableWithoutFeedback,
  Text,
  Image,
  Dimensions
} from 'react-native';

import { FloatingLabelInput, Button } from '../../components';
import { emailChanged, passwordChanged, loginUser } from './actions';
import Locales from '../../resources/locales';

import { GLOBAL } from '../../services/helpers';
import TestKeys from '../../../specs/itKeys';

const logo = require('../../resources/images/logo.png');

const SCREEN_WIDTH = Dimensions.get('window').width;

class LoginForm extends React.Component {
  onLoginButtonPress() {
    const { email, password } = this.props;

    dismissKeyboard();
    this.props.loginUser(email, password);
  }

  onEmailChange = email => this.props.emailChanged(email);
  onPasswordChange = password => this.props.passwordChanged(password);

  setInputReference = input => {
    this.passwordInput = input;
  };

  getErrorMessage() {
    return this.props.error !== '' ? this.props.translate(this.props.error) : ' ';
  }

  focusPassword = () => {
    this.passwordInput.focus();
  };

  testRef = identifier => {
    return GLOBAL.DEV ? this.props.generateTestHook(identifier) : null;
  };

  render() {
    const ButtonWrapper = wrap(Button, GLOBAL.DEV);

    return (
      <TouchableWithoutFeedback onPress={dismissKeyboard}>
        <View style={styles.container}>
          <View style={styles.imageContainer}>
            <Image style={styles.logoStyle} source={logo} resizeMode={'contain'} />
          </View>

          <FloatingLabelInput
            style={styles.emailContainer}
            ref={this.testRef(TestKeys.login.emailInput)}
            unfocusedColor={EStyleSheet.value('$colorGrey')}
            focusedColor={EStyleSheet.value('$colorPrimary')}
            autoCapitalize="none"
            autoCorrect={false}
            keyboardType="email-address"
            returnKeyType="next"
            label={this.props.translate(`${Locales.login.email}`)}
            value={this.props.email}
            onChangeText={this.onEmailChange.bind(this)}
            blurOnSubmit={false}
            onSubmitEditing={this.focusPassword.bind(this)}
          />

          <View style={styles.passwordContainer}>
            <FloatingLabelInput
              ref={this.testRef(TestKeys.login.passwordInput)}
              unfocusedColor={EStyleSheet.value('$colorGrey')}
              focusedColor={EStyleSheet.value('$colorPrimary')}
              label={this.props.translate(`${Locales.login.password}`)}
              secureTextEntry
              returnKeyType="done"
              value={this.props.password}
              withRef
              refField={this.setInputReference.bind(this)}
              onChangeText={this.onPasswordChange.bind(this)}
              blurOnSubmit
              onSubmitEditing={dismissKeyboard}
            />

            <Text ref={this.testRef(TestKeys.login.errorMessage)} style={styles.errorStyle}>
              {this.getErrorMessage()}
            </Text>
          </View>
          <ButtonWrapper
            ref={this.testRef(TestKeys.login.loginButton)}
            enabled={this.props.validEmail && this.props.validPassword}
            style={styles.loginButtonStyle}
            onPress={this.onLoginButtonPress.bind(this)}
          >
            {this.props.translate(`${Locales.login.login}`)}
          </ButtonWrapper>

          <TouchableOpacity style={styles.recoverPasswordButtonContainer}>
            <Text style={styles.recoverPasswordButtonText}>
              {this.props.translate(`${Locales.login.recover_password}`)}
            </Text>
          </TouchableOpacity>
        </View>
      </TouchableWithoutFeedback>
    );
  }
}

const styles = EStyleSheet.create({
  container: {
    flex: 1,
    paddingHorizontal: 20,
    alignItems: 'center',
    justifyContent: 'center'
  },
  imageContainer: {
    minHeight: 50,
    maxHeight: 500,
    width: SCREEN_WIDTH * 0.5,
    marginTop: 40,
    marginBottom: 50
  },
  logoStyle: {
    flex: 1,
    height: undefined,
    width: undefined
  },
  emailContainer: {
    marginBottom: 15
  },
  passwordContainer: {
    marginBottom: 40,
    alignSelf: 'stretch'
  },
  errorStyle: {
    fontSize: 11,
    marginTop: 2,
    alignSelf: 'flex-start',
    fontFamily: 'Montserrat-Regular',
    color: '$colorError'
  },
  loginButtonStyle: {
    marginBottom: 30
  },
  recoverPasswordButtonContainer: {
    marginBottom: 50
  },
  recoverPasswordButtonText: {
    fontSize: 11,
    fontFamily: 'Montserrat-Regular',
    textAlign: 'center',
    textDecorationLine: 'underline',
    color: '$colorPrimaryDark'
  }
});

const mapStateToProps = ({ login, locale }) => {
  const { email, password, validEmail, validPassword, error, isLoading } = login;

  const translate = getTranslate(locale);

  return {
    translate,
    email,
    password,
    validEmail,
    validPassword,
    error,
    isLoading
  };
};

export default connect(mapStateToProps, {
  emailChanged,
  passwordChanged,
  loginUser
})(hook(LoginForm, GLOBAL.DEV));
