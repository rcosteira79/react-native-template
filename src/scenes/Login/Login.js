import React from 'react';
import { View, Text, Image } from 'react-native';
import EStyleSheet from 'react-native-extended-stylesheet';
import LoginForm from './LoginForm';

const logo = require('../../resources/images/react-native-symbol.png');

const login = () => {
	return (
		<View style={styles.layout}>
			<View style={styles.header}>
				<Text style={styles.title}>
					Hello
				</Text>
				<View>
					<Image
						source={logo}
						style={styles.picture}
					/>
				</View>
			</View>
			<View style={styles.loginForm}>
				<LoginForm />
			</View>
		</View>
	);
};

const styles = EStyleSheet.create({
	layout: {
		backgroundColor: 'white',
		flex: 1,
	},
	header: {
		backgroundColor: '$colorAccent',
		flex: 2,
		justifyContent: 'center',
		alignItems: 'center',
	},
	title: {
		color: 'white',
		fontSize: 32,
		fontWeight: 'bold',
	},
	picture: {
		marginTop: 10,
		width: 150,
		height: 150,
	},
	loginForm: {
		flex: 3,
	},
});

export default login;
