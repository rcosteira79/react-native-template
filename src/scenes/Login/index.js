import React from 'react';
import { View } from 'react-native';
import EStyleSheet from 'react-native-extended-stylesheet';
import LoginForm from './LoginForm';
import { Card } from '../../components';

const Login = () => {
  return (
    <View style={styles.backgroundStyle}>
      <Card style={styles.cardStyle}>
        <LoginForm />
      </Card>
    </View>
  );
};

const styles = EStyleSheet.create({
  backgroundStyle: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '$colorPrimary'
  },
  cardStyle: {
    marginTop: 88,
    marginBottom: 87,
    marginLeft: 10,
    marginRight: 10
  }
});

export default Login;
