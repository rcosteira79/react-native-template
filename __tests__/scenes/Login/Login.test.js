import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Login from '../../../src/scenes/Login';

Enzyme.configure({ adapter: new Adapter() });

// Don't bother trying to DRY this. In a mock you can only require things
// locally and you aren't allowed to access external variables.
jest.mock('../../../src/navigation/routes', () => ({
  AppNavigator: {
    router: {
      getStateForAction: jest.fn(),
      getActionForPathAndParams: jest.fn()
    }
  }
}));

jest.mock('../../../src/navigation/reducer', () => ({
  AppNavigator: {
    router: {
      getStateForAction: jest.fn(),
      getActionForPathAndParams: jest.fn()
    }
  }
}));

describe('Login tests', () => {
  it('Renders the LoginScreen', () => {
    const wrapper = shallow(<Login />);
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
