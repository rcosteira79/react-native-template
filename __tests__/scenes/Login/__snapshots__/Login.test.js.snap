// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login tests Renders the LoginScreen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#55C5CE",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Card
    style={
      Object {
        "marginBottom": 87,
        "marginLeft": 10,
        "marginRight": 10,
        "marginTop": 88,
      }
    }
  >
    <Connect(LoginForm) />
  </Card>
</View>
`;
